File: NamooCalendar/src/main/java/ir/namoo/religiousprayers/praytimes/JSONUtils.java
package ir.namoo.religiousprayers.praytimes;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class JSONUtils {

    public static class City {
        public String name;
        public double lat;
        public double lng;

        @NotNull
        @Override
        public String toString() {
            return "{\"name\":" + "\"" + name + "\"," +
                    "\"lat\":" + lat + "," +
                    "\"lng\":" + lng + "}";
        }
    }

    public static class PrayTime {
        public int dayNum;
        public String fajr;
        public String sunrise;
        public String dhuhr;
        public String asr;
        public String maghrib;
        public String isha;

        @NotNull
        @Override
        public String toString() {
            return "{\"dayNum\":" + dayNum +
                    ",\"fajr\":\"" + fajr + "\"," +
                    "\"sunrise\":\"" + sunrise + "\"," +
                    "\"dhuhr\":\"" + dhuhr + "\"," +
                    "\"asr\":\"" + asr + "\"," +
                    "\"maghrib\":\"" + maghrib + "\"," +
                    "\"isha\":\"" + isha + "\"" +
                    "}";
        }
    }

    public static JSONObject toJson(City city, List<PrayTime> prayTimes) {
        if (city == null || prayTimes == null)
            return null;
        JSONObject res = new JSONObject();
        res.put("city", city);
        JSONArray plist = new JSONArray();
        plist.addAll(prayTimes);
        res.put("times", plist);
        return res;
    }

    @Nullable
    public static City getCity(String strJson) {
        try {
            JSONParser jsonParser = new JSONParser();
            Object obj = jsonParser.parse(strJson);
            JSONObject jsonObject = (JSONObject) obj;
            Object jsonCity = jsonObject.get("city");
            JSONObject jCity = (JSONObject) jsonCity;
            if (jCity == null) return null;
            City res = new City();
            res.name = Objects.requireNonNull(jCity.get("name")).toString();
            res.lat = Double.parseDouble(Objects.requireNonNull(jCity.get("lat")).toString());
            res.lng = Double.parseDouble(Objects.requireNonNull(jCity.get("lng")).toString());
            return res;
        } catch (Exception ex) {
            System.out.println("Error: " + ex);
            ex.printStackTrace();
            return null;
        }
    }

    @Nullable
    public static List<PrayTime> getPrayTimes(String strJson) {
        try {
            List<PrayTime> res = new ArrayList<>();
            JSONParser jsonParser = new JSONParser();
            Object obj = jsonParser.parse(strJson);
            JSONObject jsonObject = (JSONObject) obj;
            Object jsonTimes = jsonObject.get("times");
            JSONArray times = (JSONArray) jsonTimes;
            if (times == null) return null;
            for (Object j : times) {
                JSONObject t = (JSONObject) j;
                PrayTime temp = new PrayTime();
                temp.dayNum = Integer.parseInt(Objects.requireNonNull(t.get("dayNum")).toString());
                temp.fajr = (String) t.get("fajr");
                temp.sunrise = (String) t.get("sunrise");
                temp.dhuhr = (String) t.get("dhuhr");
                temp.asr = (String) t.get("asr");
                temp.maghrib = (String) t.get("maghrib");
                temp.isha = (String) t.get("isha");
                res.add(temp);
            }
            return res;
        } catch (Exception ex) {
            System.out.println("Error: " + ex);
            ex.printStackTrace();
            return null;
        }
    }

}//end of JSONUtils

File: NamooCalendar/src/main/java/ir/namoo/religiousprayers/praytimes/JSONUtils.kt
package ir.namoo.religiousprayers.praytimes

import org.json.simple.JSONArray
import org.json.simple.JSONObject
import org.json.simple.parser.JSONParser
import java.util.*

class JSONCity {
    var name: String? = null
    var lat = 0.0
    var lng = 0.0
    override fun toString(): String {
        return "{\"name\":" + "\"" + name + "\"," +
                "\"lat\":" + lat + "," +
                "\"lng\":" + lng + "}"
    }
}

class JSONPrayTime {
    var dayNum = 0
    var fajr: String? = null
    var sunrise: String? = null
    var dhuhr: String? = null
    var asr: String? = null
    var maghrib: String? = null
    var isha: String? = null
    override fun toString(): String {
        return "{\"dayNum\":" + dayNum +
                ",\"fajr\":\"" + fajr + "\"," +
                "\"sunrise\":\"" + sunrise + "\"," +
                "\"dhuhr\":\"" + dhuhr + "\"," +
                "\"asr\":\"" + asr + "\"," +
                "\"maghrib\":\"" + maghrib + "\"," +
                "\"isha\":\"" + isha + "\"" +
                "}"
    }
}

fun toJson(
    city: JSONCity?,
    prayTimes: List<JSONPrayTime?>?
): JSONObject? {
    if (city == null || prayTimes == null) return null
    val res = JSONObject()
    res["city"] = city
    val plist = JSONArray()
    plist.addAll(prayTimes)
    res["times"] = plist
    return res
}

fun getCity(strJson: String?): JSONCity? {
    return try {
        val jsonParser = JSONParser()
        val obj = jsonParser.parse(strJson)
        val jsonObject = obj as JSONObject
        val jsonCity = jsonObject["city"]
        val jCity = jsonCity as JSONObject? ?: return null
        val res = JSONCity()
        res.name = Objects.requireNonNull<Any>(jCity["name"]).toString()
        res.lat =
            Objects.requireNonNull<Any>(jCity["lat"]).toString().toDouble()
        res.lng =
            Objects.requireNonNull<Any>(jCity["lng"]).toString().toDouble()
        res
    } catch (ex: Exception) {
        println("Error: $ex")
        ex.printStackTrace()
        null
    }
}

fun getPrayTimes(strJson: String?): List<DownloadedPrayTimesEntity>? {
    return try {
        val res: MutableList<DownloadedPrayTimesEntity> =
            ArrayList()
        val jsonParser = JSONParser()
        val obj = jsonParser.parse(strJson)
        val jsonObject = obj as JSONObject
        val jsonTimes = jsonObject["times"]
        val times = jsonTimes as JSONArray? ?: return null
        for (j in times) {
            val t = j as JSONObject
            val temp = DownloadedPrayTimesEntity(
                id = 0,
                city = getCity(strJson)!!.name!!,
                dayNumber = t["dayNum"].toString().toInt(),
                fajr = t["fajr"] as String,
                sunrise = t["sunrise"] as String,
                dhuhr = t["dhuhr"] as String,
                asr = t["asr"] as String,
                maghrib = t["maghrib"] as String,
                isha = t["isha"] as String
            )

            res.add(temp)
        }
        res
    } catch (ex: java.lang.Exception) {
        println("Error: $ex")
        ex.printStackTrace()
        null
    }
}
